mss/L2Cache.cpp

```diff
@@ -0,0 +1,657 @@
+            biu_req_queue_.emplace_back(inst_ptr);
+
+            ev_handle_l2cache_biu_req_.schedule(sparta::Clock::Cycle(0));
+        }
+        else {
+            sparta_assert("No space in biu_req_queue_! Why did the frontend issue push the request onto l2cache_pipeline_?");
+        }
+        
+        ILOG("Append L2Cache->BIU req queue");
+    }
+    
+    // Return the resp to the master units
+    void L2Cache::sendOutResp_(const L2ArchUnit &unit, const olympia::InstPtr& instPtr) {
+        // if (instPtr is originally from DCache)
+        if (unit == L2ArchUnit::DCACHE) {
+            appendDCacheRespQueue_(instPtr);
+        }
+        // if (instPtr is originally from ICache)
+        else if (unit == L2ArchUnit::ICACHE) {
+            appendICacheRespQueue_(instPtr);
+        }
+        else {
+            sparta_assert("Resp is being sent to a Unit that is not valid");
+        }
+    }
+
+    // Send the request to the slave units
+    void L2Cache::sendOutReq_(const L2ArchUnit &unit, const olympia::InstPtr& instPtr) {
+        // if (instPtr is destined for BIU on L2Cache miss) 
+        if (unit == L2ArchUnit::BIU) {
+            appendBIUReqQueue_(instPtr);
+        }
+        else {
+            sparta_assert("Request is being sent to a Unit that is not valid");
+        }
+    }
+
+    // Select the Channel to pick the request from
+    // Current options : 
+    //       BIU - P0
+    //       DCache - P1 - RoundRobin Candidate
+    //       DL1 - P1 - RoundRobin Candidate
+    L2Cache::Channel L2Cache::arbitrateL2CacheAccessReqs_() {
+        sparta_assert(icache_req_queue_.size() > 0 || dcache_req_queue_.size() > 0 || biu_resp_queue_.size() > 0,
+                                "Arbitration failed: Reqest queues are empty!");
+        
+        Channel winner;
+
+        if (pipeline_req_queue_.numFree() == 0) {
+
+            // pipeline_req_queue_ is full
+            return Channel::NO_ACCESS;
+        }
+
+        // P0 priority to sevice the pending request in the buffer
+        if (!biu_resp_queue_.empty()) {
+            winner = Channel::BIU;
+            ILOG("Arbitration winner - BIU");
+
+            return winner;
+        }
+        
+        // RoundRobin for P1 Priority
+        while (true) {
+            
+            if (channel_select_ == Channel::ICACHE) {
+                channel_select_ = Channel::DCACHE;
+                if (!icache_req_queue_.empty()) {
+                    winner = Channel::ICACHE;
+                    ILOG("Arbitration winner - ICache");
+                    
+                    return winner;
+                }
+            }
+            else if (channel_select_ == Channel::DCACHE) {
+                channel_select_ = Channel::ICACHE;
+                if (!dcache_req_queue_.empty()) {
+                    winner = Channel::DCACHE;
+                    ILOG("Arbitration winner - DCache");
+
+                    return winner;
+                }
+            }
+        }
```

Comment:
Why is this in a `while(true)` loop?   There seems to be no iterations here -- and what happens if the `channel_select_` isn't one of I/Dcache?

Reply:
I think this was a remnant of the previous implementation. I have fixed this in my local change. Will push it soon. Thanks for pointing it out. :)